Run-2644: Functional tests for job activity history

jayas006 opened the pull request:
**Is this a bugfix, or an enhancement? Please describe.**

This is an enhancement. 

**Describe the solution you've implemented**
	I have added Selenium functional tests to check job activity history on the following pages:
             **Job List Page:**
	•	Ensures job activity history is retrieved correctly.
               **Job List Page with Saved Filters:**
	•	Confirms filters are saved.
                 **Activity Page:**
	•	Checks activity logs are displayed properly

Additionally, I implemented a helper method to simplify the tests by reducing repetitive code. This method handles common actions like finding elements and waiting for them to load, making the tests easier to read and maintain
 
**Describe alternatives you've considered**
<!-- A clear and concise description of any alternative solutions or features you've considered. -->


**Additional context**
<!-- Add any other context or screenshots about the change here. -->

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/api/job/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,54 @@
+
+    def "Retrieve job Activity history from Job List Page "() {
+        given: "A Job exists in the project with execution history"
+        def jobId = "3957f735-5269-4c85-bc57-f123598886f6"
```
Why would this job exist?

jayas006 replied:
> Why would this job exist?

I have used this jobID to represent a job that already exists in the project. This makes it easier to test if the API can get the job's details.  Do I need to create a job first before checking its activity?

mrdubr replied:
Your tests should be independent. Given a test suite TS with (optional) setup and tear down functions and multiple tests t1, t2, ..., tn in that suite. You should be able to run `tx` at any time and in any order. 
If you deviate from this rule, there should be a very good reason.

Thus, if your test relies on an job guide, it should be created by the test itself (or setup functionality).

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/api/job/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,54 @@
+
+        when: "The activity history for the job is requested"
+
+        def response = doGet("/job/${jobId}/executions")
```
Use org.rundeck.util.common.jobs.JobUtils functions to retrieve job data. This is the layer where the REST API calls are meant to be isolated. Avoid calling the API directly from the tests.

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/api/job/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,54 @@
+                    executionMap.containsKey(key)
+
+                }
+            }
```
Keep in mind the AC of your tickets.
https://pagerduty.atlassian.net/browse/RUN-2644?atlOrigin=eyJpIjoiN2YwM2VkYzcxMWM5NGVhNDg4NWUwNTUzNzk0NjBiMGIiLCJwIjoiaiJ9

You're writing (UI) "selenium" test that very "job history"

jayas006 replied:
> Keep in mind the AC of your tickets. https://pagerduty.atlassian.net/browse/RUN-2644?atlOrigin=eyJpIjoiN2YwM2VkYzcxMWM5NGVhNDg4NWUwNTUzNzk0NjBiMGIiLCJwIjoiaiJ9
> 
> You're writing (UI) "selenium" test that very "job history"

Thanks, I understand that I need to follow selenium based test structure for this. I am reviewing the existing selenium tests for guidance and update my code accordingly. It's great to see the browser interaction during the tests.

mrdubr reviewed (changes requested):
Good start!
Please, see my comments.
Keep on working on this PR!

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,82 @@
+        loginPage.login(TEST_USER, TEST_PASS)
+        println "Login successful"
+    }
+    def "create job has basic fields"() {
```
this test is not related to your ticket AC

jayas006 replied:
> this test is not related to your ticket AC

Thank you for pointing that out. I initially thought the test needed to include the flow of creating a job and then executing it. I have removed it

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,82 @@
+        println "Navigating to Job Show Page"
+        jobShowPage.validatePage()
+        println "Job Show Page validated"
+        def runJobButton = driver.findElement(By.cssSelector(".btn.btn-success.btn-simple.btn-hover.btn-xs.act_execute_job"))
```
Isolate css selectors to the `JobShowPage`.
Use "helpers" from the `JobShowPage` e.g. `runJobLink`

jayas006 replied:
> Isolate css selectors to the `JobShowPage`. Use "helpers" from the `JobShowPage` e.g. `runJobLink`

I have not  implemented  this yet as I wanted to confirm it first.

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,82 @@
+        requestOptionField.sendKeys("test-value")
+        println "Filled required option field"
+
+        def createJobButton = driver.findElement(By.id("execFormRunButton"))
```
is it `create` or `run`?

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,82 @@
+        jobCreatePage.descriptionTextarea
+        println "job created successfully"
+    }
+    def "run job and validate execution"() {
```
this test is not related to your ticket AC

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,82 @@
+import org.rundeck.util.gui.pages.jobs.JobShowPage
+import spock.lang.Stepwise
+@SeleniumCoreTest
+@Stepwise
```
You should avoid introducing an execution sequence to the tests as it is generally considered a bad practice in unit testing. Rewrite your tests by factoring out shared logic to "setup" and "setupspec".

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,82 @@
+        executionShowPage.validateStatus('SUCCEEDED')
+        println "Job executed successfully"
+    }
+    def "validate activity history from Job List Page"() {
```
I've added screenshots to the ticket. 
Please, ensure you cover all scenarios in the AC.

mrdubr reviewed (changes requested):
Please, see my comments.

jayas006 commented on functional-test/src/test/groovy/org/rundeck/util/gui/pages/jobs/JobShowPage.groovy:
```diff
@@ -48,6 +48,8 @@ class JobShowPage extends BasePage implements ActivityListTrait {
     By jobLinkTitleBy = By.xpath("//a[contains(@class, 'job-header-link')]")
     By autocompleteJobStepDefinitionBy = By.cssSelector("#wfitem_0 > span > div > div > span > span > span.text-success")
     By runFormBy = By.cssSelector("#execDiv #exec_options_form #formbuttons #execFormRunButton")
+    By runJobButtonBy = By.cssSelector(".btn.btn-success.btn-simple.btn-hover.btn-xs.act_execute_job")
+    By executionStateLabelBy = By.id("executionStateDisplayLabel")
```
@mrdubr  before using it in my test I wanted to make sure if I am implementing it correctly?  Do I need to use these selector with the helper method clickRunJobButton and waitForExecutionState

mrdubr replied:
IIRC the ticket AC, your tests are not related to the JobShow page, thus I'm not sure you'd need this.
It's much simpler (and more maintainable) to setup (prepare) data for your tests using backend API.
There are numerous tests that create jobs and wait for executions to complete.
I'd recommend following that pattern.

jayas006 replied:
> IIRC the ticket AC, your tests are not related to the JobShow page, thus I'm not sure you'd need this. It's much simpler (and more maintainable) to setup (prepare) data for your tests using backend API. There are numerous tests that create jobs and wait for executions to complete. I'd recommend following that pattern.

I tried to refactor the tests by looking other tests. Though it's failing but I am keep working on it. It  doesn't look complicated but I may be missing something.

mrdubr replied:
Feel free to zoom me if you need some guidance on this

jayas006 commented:
@mrdubr I have refactored the tests and now each test is independent. The tests look smaller and more concise. I have declared the CSS selector in the test file temporarily to check if they works. Since I have confirmed they are  working, I will move them to the Execution and Job page as  helper methods. I understand that adding them as helper method is Better for clarity and reusability.

mrdubr commented on functional-test/src/test/groovy/org/rundeck/util/gui/pages/jobs/JobListPage.groovy:
```diff
@@ -52,7 +52,49 @@ class JobListPage extends BasePage implements ActivityListTrait {
+    static final By SAVED_FILTER_LINK = By.cssSelector("a[data-test='filter-link']")
+    static final By ACTIVITY_ROWS = By.cssSelector("table.activity-list-table tbody tr.link.activity_row")
+    static final By EVENT_ICON = By.cssSelector("td.eventicon")
+/** Click Save Filter button */
```
We tend to stick to the javadoc commenting style: https://www.oracle.com/ca-en/technical-resources/articles/java/javadoc-tool.html

In particular, in this case, the indentation is unconventional.

mrdubr commented on functional-test/src/test/groovy/org/rundeck/util/gui/pages/jobs/JobShowPage.groovy:
```diff
@@ -136,10 +136,12 @@ class JobShowPage extends BasePage implements ActivityListTrait {
         }
     }
 
+
```
Please, avoid introducing unnecessary formatting changes. It adds noise to a PR w/o providing value.

mrdubr replied:
Revert these unneeded changes

mrdubr commented on functional-test/src/test/groovy/org/rundeck/util/gui/pages/jobs/JobListPage.groovy:
```diff
@@ -52,7 +52,49 @@ class JobListPage extends BasePage implements ActivityListTrait {
+        el(SAVE_FILTER_BUTTON).click()
+        return this
+    }
+/** Enter filter name */
```
Also, notice how in the multiple methods above the comment itself does not add any value since it's equivalent to the method name. This makes the comment unnecessary.

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    def setup() {
+        wait = new WebDriverWait(getDriver(), Duration.ofSeconds(30))
```
Is there a reason you're configuring/reconfiguring `wait`. If so, please add a comment explaining that.

jayas006 replied:
> Is there a reason you're configuring/reconfiguring `wait`. If so, please add a comment explaining that.

I have explained it in the comment, I added it because the test was failing in CircleCI due to element took longer and it was passing locally. Increasing the wait duration resolved the issue.

mrdubr replied:
I see... You're using this to wait for the job to get into the `SUCCEEDED` status.

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,106 @@
+
+    def setup() {
+        wait = new WebDriverWait(getDriver(), Duration.ofSeconds(30))
+        // Login
```
Useless comment

jayas006 replied:
> Comment that does not provide additional value

Thanks, I have removed all the redundant comments.

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,106 @@
+        loginPage.login(TEST_USER, TEST_PASS)
+        // Create a job
+        def jobDefinition = JobUtils.generateScheduledExecutionXml("TestJobActivityHistory")
+        def client = getClient() // Assuming a method exists to get RdClient
```
Does this comment add value?

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,106 @@
+        def response = JobUtils.createJob(SELENIUM_BASIC_PROJECT, jobDefinition, client)
+        assert response.succeeded.size() > 0: "Failed to create job"
+        jobId = response.succeeded[0].id
+        println "Job created with ID: ${jobId}"
```
Please, ensure that debugging statements and any temporary code are removed from your code before a code review is started.

jayas006 replied:
> Please, ensure that debugging statements and any temporary code are removed from your code before a code review is started.

Yes I have checked few times and removed the debugging statements.

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    def cleanup() {
+        getDriver().quit()
```
Is there a reason you need this? If so, please add a comment explaining that.

jayas006 replied:
> Is there a reason you need this? If so, please add a comment explaining that.

It was needed earlier when my previous script caused test failures in CircleCI but since updating the script, I have removed it as well

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    /**
+     Test 1: Review Job Activity History from Job List Page
```
Your comment is equivalent to the test name and does not provide any additional value.

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,106 @@
+        WebElement statusIcon = ExecutionShowPage.getActivityExecStatusIcon(firstActivityRow)
+        String status = statusIcon.getAttribute("data-statusstring")
+        println "First Activity Row Status: ${status}"
+        assert status.equalsIgnoreCase("SUCCEEDED"): "Expected 'SUCCEEDED', but found: '${status}'"
```
What will ensure that the job has finished executing by the time this assertion is checked?

mrdubr replied:
I see now that the `wait.until` on line 56 will ensure this! 👍

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,94 @@
+        jobListPage.openFilterDropdown()
+                .selectSavedFilter()
+        then: "Validate that the saved filter is applied and results are filtered"
+        def filteredRows = jobListPage.getActivityRows()
```
The saved/selected filter is the same as the default 'any time' filter, thus this test does not verify that filtering applies.

jayas006 replied:
> The saved/selected filter is the same as the default 'any time' filter, thus this test does not verify that filtering applies.

I didn't understand this. Just for clarification,  I need to update the test to ensure it applies specific filter criteria ("In Last day/ In last week") when saving the filter and validate that all the rows in the filtered results meet the criteria?

mrdubr replied:
You need to determine if verifying **correctness** of a filter application is in scope for your work. 
If not, state this explicitly. If it's not possible to do so (e.g. due to test data provisioning constraints) -- state that explicitly. 

Otherwise, it looks like this test is testing filtering correctness, but, in reality, it only tests the `*` (any time) fitler.

jayas006 replied:
> You need to determine if verifying **correctness** of a filter application is in scope for your work. If not, state this explicitly. If it's not possible to do so (e.g. due to test data provisioning constraints) -- state that explicitly.
> 
> Otherwise, it looks like this test is testing filtering correctness, but, in reality, it only tests the `*` (any time (aka lack of filtering)) filter.

Thank you for your feedback. I have made changes `in the last week` instead of` anytime` to improve this test. using anytime filter, I am now testing with a specific time filter. I appreciate your guidance, as it is helping me make my tests more accurate. I am learning a lot and striving to perfect my work.

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,94 @@
+        def firstActivityRow = activityList.get(0)
+        def firstActivityStatusIcon = ExecutionShowPage.getActivityExecStatusIcon(firstActivityRow)
+        def firstActivityStatus = firstActivityStatusIcon.getAttribute("data-statusstring")
+        assert firstActivityStatus.equalsIgnoreCase("SUCCEEDED"):
```
Ditto

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,94 @@
+        def filteredRows = jobListPage.getActivityRows()
+        assert filteredRows.size() > 0: "No filtered results found"
+        def firstRowStatus = jobListPage.getFirstRowStatus()
+        assert firstRowStatus.equalsIgnoreCase('SUCCEEDED'):
```
Ditto

mrdubr reviewed (changes requested):
Great progress!
Please, see my comments for the additional info on the improvement recommendations.

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,94 @@
+    def setup() {
+        // Set up WebDriverWait to wait up to 30 seconds for elements that take time to appear or change
+        wait = new WebDriverWait(getDriver(), Duration.ofSeconds(30))
+        def loginPage = go LoginPage
```
nit:

Note: My `nit` comments are FYIs, subjective and will not stop the PR from getting approved! :-) 

In general, it's considered too noisy to introduce a var to use it *once* right away.
This could be compressed to something like `(go LoginPage).login(TEST_USER, TEST_PASS)`

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,94 @@
+        then: "Validate job execution is listed in Activity History"
+        activityPage.validatePage()
+        List<WebElement> activityList = activityPage.getActivityRows()
+        assert activityList.size() > 0: "No activity rows found"
```
```suggestion
        !activityList.empty()
```
In general, avoid using `size()` to check for emptiness for perf reasons

Also, in a Spock specification, you do not necessarily need to use `assert` statements in the `then` section. Spock's then block is designed for assertions, and it implicitly treats any boolean expression as an assertion.

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,94 @@
+        assert activityList.size() > 0: "No activity rows found"
+        WebElement firstActivityRow = activityList.get(0)
+        WebElement statusIcon = ExecutionShowPage.getActivityExecStatusIcon(firstActivityRow)
+        wait.until{statusIcon.getAttribute("data-statusstring").equalsIgnoreCase("SUCCEEDED")}
```
You're using UI state to wait for the backend job to succeed. Technically, it's correct but slightly unconventional. I'd add a comment explaining this to remove ambiguity.

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,94 @@
+        WebElement firstActivityRow = activityList.get(0)
+        WebElement statusIcon = ExecutionShowPage.getActivityExecStatusIcon(firstActivityRow)
+        wait.until{statusIcon.getAttribute("data-statusstring").equalsIgnoreCase("SUCCEEDED")}
+        String status = statusIcon.getAttribute("data-statusstring")
+        assert status.equalsIgnoreCase("SUCCEEDED"): "Expected 'SUCCEEDED', but found: '${status}'"
```
I'd assume `wait.until` ensure this is true. I don't think we need these lines.

mrdubr reviewed (changes requested):
You're on the write path! 👍 
Just need a few tweaks to make this test suite resilient!

jayas006 commented:
> You're on the write path! 👍 Just need a few tweaks to make this test suite resilient!

@mrdubr, Thank you so much for your feedback. I believe I have addressed all your comments and made improvements to the test. There might be few recent changes left to review. I plan to go over it a few more times to ensure nothing is missing. Once that’s done, I think the test will be ready for review. I truly appreciate your help in making this test as perfect as possible.

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,97 @@
+        assert response.succeeded.size() > 0: "Failed to create job"
+        jobId = response.succeeded[0].id
+        def runResponse = JobUtils.executeJob(jobId, client)
+        assert runResponse.code() == 200: "Failed to execute job"
```
Using assert within the setup method is generally not recommended because setup is intended for initializing the test environment rather than for performing assertions. Assertions should be placed within test methods to validate conditions.

jayas006 replied:
> nit: Using assert within the setup method is generally not recommended because setup is intended for initializing the test environment rather than for performing assertions. Assertions should be placed within test methods to validate conditions.

So  I remove the assertions and use throw exceptions if initialization fails?

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,97 @@
+        then: "Validate job execution is listed in Activity History"
+        activityPage.validatePage()
+        wait.until {
+            def activityList = activityPage.getActivityRows()
```
nit:
ditto
`!activityPage.getActivityRows().isEmpty()`

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,97 @@
+            !activityList.isEmpty()
+        }
+
+        def firstActivityRow = activityPage.getActivityRows().get(0)
```
Unnecessary validation:
Your test does not need to check if the execution has succeeded. 
You are testing "activity", not job execution status. Seeing a row is sufficient.
Contrary to the initial gut feel, "less is more" with unit testing. Keep them focused.

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,97 @@
+    def "validate activity history from Job List Page Saved Filters"() {
+        when: "Navigate to the Job List Page and apply a saved filter"
+        def jobListPage = go JobListPage, SELENIUM_BASIC_PROJECT
+        jobListPage.clickAnyTimeButton()
```
This code block is so elegant! 👍
Compact, declarative, to the point! 🥇

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,97 @@
+        then: "Wait for the job to complete and validate the saved filter is applied"
+        wait.until {
+            def status = jobListPage.getFirstRowStatus()
+            status.equalsIgnoreCase('SUCCEEDED')
```
ditto

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,97 @@
+            status.equalsIgnoreCase('SUCCEEDED')
+        }
+
+        and: "Validate that the results are filtered"
```
I'm not sure I understand how this confirms that the results are filtered. What value does this assertion add to the one above?

jayas006 replied:
> I'm not sure I understand how this confirms that the results are filtered. What value does this assertion add to the one above?

The goal  of this check is to make sure the saved filter is actually applied. While the previous check ensuring the job status is succeeded. I thought by adding this it ensures that the filtered results are not empty and match the filter criteria.  Do you have suggestions for a better way to confirm the filter is working? If I missing anything please let me know, I will be happy to revise it

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    def "review job activity history from Activity Page"() {
+        when: "Navigate to the Activity Page"
```
don't you validate the same things in the `"review job activity history from Job List Page"` test already?

jayas006 replied:
> don't you validate the same things in the `"review job activity history from Job List Page"` test already?

Yes you are correct. This test and review job activity history from job list page test do validate similar functionality. The main intent behind this test was to ensure that activity history is accessible and validated directly from the Activity Page. I will revise this test and remove it if redundant. Thank you for pointing this out

jayas006 replied:
> > don't you validate the same things in the `"review job activity history from Job List Page"` test already?
> 
> Yes you are correct. This test and review job activity history from job list page test do validate similar functionality. The main intent behind this test was to ensure that activity history is accessible and validated directly from the Activity Page. I will revise this test and remove it if redundant. Thank you for pointing this out

Also it required according to the ticket `Review Job Activity History From Activity Page.`

mrdubr replied:
Please, take a look at the screenshots in: https://pagerduty.atlassian.net/browse/RUN-2644?atlOrigin=eyJpIjoiZDgxYzVjZjAyMmFmNDQ1M2I4NTZhNThhMDgwZmFjN2YiLCJwIjoiaiJ9

The "job activity" appears on **two** different pages.
It seems that the tests are only testing one of the pages.

jayas006 replied:
Yes both pages are validating to activity instead one of them validate to Job page where JobLists are. So for the first test it should show only to Job Page and third test for activity page. Thanks for pointing that out

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,97 @@
+        activityPage.loadActivityPageForProject(SELENIUM_BASIC_PROJECT)
+
+        then: "Validate job execution is listed in Activity History"
+        activityPage.validatePage()
```
Already validated above

mrdubr commented on functional-test/src/test/groovy/org/rundeck/util/gui/pages/activity/ActivityPage.groovy:
```diff
@@ -44,4 +45,11 @@ class ActivityPage extends BasePage implements ActivityListTrait{
     }
 
+    static void validateFirstActivityRow(List<WebElement> activityList) {
+        assert !activityList.isEmpty() : "Activity list is empty, no rows to validate."
```
Assertions are for tests, but this is the "model".
You're using groovy assertions here (that are different from Spec assertions) that are generally used to asserts invariants -- e.g. null checks.

mrdubr commented on functional-test/src/test/groovy/org/rundeck/util/gui/pages/activity/ActivityPage.groovy:
```diff
@@ -44,4 +45,11 @@ class ActivityPage extends BasePage implements ActivityListTrait{
+        WebElement firstActivityRow = activityList.get(0)
+        WebElement statusIcon = ExecutionShowPage.getActivityExecStatusIcon(firstActivityRow)
+        String status = statusIcon.getAttribute("data-statusstring")
+        assert status.equalsIgnoreCase("SUCCEEDED") : "Expected 'SUCCEEDED', but found: '${status}'"
```
ditto: activity is different from "success".

mrdubr commented on functional-test/src/test/groovy/org/rundeck/util/gui/pages/jobs/JobListPage.groovy:
```diff
@@ -52,7 +52,61 @@ class JobListPage extends BasePage implements ActivityListTrait {
+
+    JobListPage clickLastWeekButton() {
+        waitForElementVisible(IN_THE_LAST_WEEK_BUTTON).click()
+        return this
```
I don't think you need the `return` statement.

jayas006 replied:
> I don't think you need the `return` statement.

the return this statements are necessary to support method chaining in the test casesWithout the return this statement, chaining multiple actions like 
clickAnyTimeButton()
             ```
   .clickLastWeekButton()
                .clickSaveFilterButton()
                .enterFilterName("testFilter")
                .confirmFilterSave()
                .openFilterDropdown()
                .selectSavedFilter()
```
                would not work. 
                If I remove the return this then I have to refactor my script like this 
              ```
  jobListPage.clickAnyTimeButton()
                jobListPage .clickLastWeekButton()
                jobListPage  .clickSaveFilterButton()
```
        and test is passing

jayas006 replied:
> > I don't think you need the `return` statement.
> 
> the return this statements are necessary to support method chaining in the test casesWithout the return this statement, chaining multiple actions like
> 
> ```
> clickAnyTimeButton()
>    .clickLastWeekButton()
>                 .clickSaveFilterButton()
>                 .enterFilterName("testFilter")
>                 .confirmFilterSave()
>                 .openFilterDropdown()
>                 .selectSavedFilter()
> ```

But after removing return this I need to refactor test like this 
              ```
  jobListPage.clickAnyTimeButton()
                jobListPage .clickLastWeekButton()
                jobListPage  .clickSaveFilterButton()

```

mrdubr replied:
Good point! Thanks for the clarification!

mrdubr commented on functional-test/src/test/groovy/org/rundeck/util/gui/pages/jobs/JobListPage.groovy:
```diff
@@ -52,7 +52,61 @@ class JobListPage extends BasePage implements ActivityListTrait {
+    }
+    JobListPage clickSaveFilterButton() {
+        waitForElementVisible(SAVE_FILTER_BUTTON).click()
+        return this
```
ditto

mrdubr reviewed (changes requested):
Good progress! 👍 
Please see my comments.

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -30,14 +29,16 @@ class JobActivityHistorySpec extends SeleniumBase {
         def client = getClient()
         def response = JobUtils.createJob(SELENIUM_BASIC_PROJECT, jobDefinition, client)
-        assert response.succeeded.size() > 0: "Failed to create job"
+        if (response.succeeded.size() == 0) {
+            throw new RuntimeException("Failed to create job")
+        }
```
There is a fine line between defensive programming and unnecessary code. 
Are you sure you need to check the `response`? What's the error behaviour of the helper function you're calling? That's what my comment was alluding to. 

Groovy assertion (don't confuse them with Spec assertions) are perfectly appropriate **when needed** and are definitely prefered over `if(error_condition) { throw }` due to compactness and intention clarity.

mrdubr commented on functional-test/src/test/groovy/org/rundeck/util/gui/pages/activity/ActivityPage.groovy:
```diff
@@ -44,4 +45,10 @@ class ActivityPage extends BasePage implements ActivityListTrait{
         els By.xpath("//*[contains(text(),'${jobName}')]")
     }
 
+    static void validateFirstActivityRow(List<WebElement> activityList) {
```
the term `validate` in the name of this function does not convey much meaning. What does it mean for the first activity raw to be valid or invalid? It depends on the context.

Also, `Exception`s (albeit overused in java) are reserved for something exceptional.
Having no activity is a perfectly valid state in our business logic. A boolean would make much more sense.

E.g. function along the lines of `boolean hasActivity(List<WebElement> activityList)`

Now, we can simplify this further employing groovy truthy logic
```
!!activityList
```

So, do we even need this function?

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,82 @@
+        when: "Navigate to the Job List Page"
+        def activityPage = go ActivityPage, SELENIUM_BASIC_PROJECT
+        activityPage.loadActivityPageForProject(SELENIUM_BASIC_PROJECT)
+        println "Navigated to Activity Page"
```
ditto

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,82 @@
+        setupProject(SELENIUM_BASIC_PROJECT)
+    }
+
+    /**
+     * Setup the test environment.
+     */
```
nit: adds no value
```suggestion
```

mrdubr reviewed (changes requested):
Almost there! 👍 
A few more things to take care of!

jayas006 commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,73 @@
+                .confirmFilterSave()
+                .openFilterDropdown()
+                .selectSavedFilter()
+        then: "Validate the saved filter is applied"
+        wait.until{
+            jobListPage.getAppliedFilterName() == "TestFilter"
+        }
+
+        }
+
```
I changed the validation to check if the saved filter is applied by comparing it to the filter name TestFilter. This is different from the first test, which checks if job activity rows are present. For this second test, the goal is to make sure the saved filter is correctly applied when viewing job activity history

mrdubr replied:
Are you validating that the filtering logic is correct -- only activity matching the updated filter shows up? If not, the test name does not make much sense: `"validate activity history from Job List Page Saved Filters"`

jayas006 commented:
> Almost there! 👍 A few more things to take care of!

Thanks for your feedback. I have tried to applied all the pointers. Double checked if I left any debugging statements.

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,73 @@
+            jobListPage.getAppliedFilterName() == "TestFilter"
+        }
+
+        }
```
white spacing

mrdubr commented on functional-test/src/test/groovy/org/rundeck/util/gui/pages/activity/ActivityPage.groovy:
```diff
@@ -44,4 +45,8 @@ class ActivityPage extends BasePage implements ActivityListTrait{
         els By.xpath("//*[contains(text(),'${jobName}')]")
     }
 
+    static boolean hasActivity(List<WebElement> activityList) {
```
you do not need this static function anymore in this class anymore

jayas006 commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -34,16 +34,16 @@ class JobActivityHistorySpec extends SeleniumBase {
     def "review job activity history from Job List Page"() {
         when: "Navigate to the Job List Page"
         def JobListPage = go JobListPage, SELENIUM_BASIC_PROJECT
-        JobListPage.loadJobListForProject(SELENIUM_BASIC_PROJECT)
```
I removed the `loadJobListForProject` method because the `go JobListPage` method already handling navigation and set up the page context. I believe  loadJobListForProject was not adding any additional value and the test works consistently without it.

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,74 @@
+        then: "Validate job execution is listed in Activity History"
+        JobListPage.validatePage()
+        wait.until {
+        !JobListPage.getActivityRows().isEmpty()
+
```
indentation

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,74 @@
+
+    def "review job activity history from Job List Page"() {
+        when: "Navigate to the Job List Page"
+        def JobListPage = go JobListPage, SELENIUM_BASIC_PROJECT
```
Variable name.
Are you using IntelliJ? It should be giving you warning for case violations.
Code review is not the appropriate time to detect such errors.

jayas006 replied:
> Variable name. Are you using IntelliJ? It should be giving you warning for case violations. Code review is not the appropriate time to catch such errors.

Yes right, I know what may happened. When I copied the class name and used it as a variable then missed to use the defined variable name. But yes I didn't get any warning in intelliJ

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,74 @@
+        activityPage.loadActivityPageForProject(SELENIUM_BASIC_PROJECT)
+
+        then: "Validate job execution is listed in Activity History"
+        !activityPage.getActivityRows().isEmpty()
```
How can we be sure that the job has started (has some activity) prior to running this check?

jayas006 replied:
Need to use wait.until ?

mrdubr commented on functional-test/src/test/groovy/org/rundeck/util/gui/pages/activity/ActivityPage.groovy:
```diff
@@ -5,6 +5,7 @@ import org.openqa.selenium.By
 import org.openqa.selenium.WebElement
 import org.rundeck.util.container.SeleniumContext
 import org.rundeck.util.gui.pages.BasePage
+import org.rundeck.util.gui.pages.execution.ExecutionShowPage
```
do we need this import?

mrdubr commented on functional-test/src/test/groovy/org/rundeck/util/gui/pages/execution/ExecutionShowPage.groovy:
```diff
@@ -105,6 +107,11 @@ class ExecutionShowPage extends BasePage  implements ActivityListTrait{
         el optionValueSelected
     }
 
+    static WebElement getActivityExecStatusIcon(WebElement row) {
```
This static method is not safe and hard to reuse.
It takes a generic WebElement and returns a generic WebElement.
What stops someone from passing in a WebElement that has nothing to do with the row?

mrdubr commented on functional-test/src/test/groovy/org/rundeck/util/gui/pages/execution/ExecutionShowPage.groovy:
```diff
@@ -2,6 +2,7 @@ package org.rundeck.util.gui.pages.execution
 
 import groovy.transform.CompileStatic
 import org.openqa.selenium.By
+
```
don't introduce unneeded formatting changes

mrdubr commented on rundeckapp/grails-spa/packages/ui-trellis/src/app/utilities/locales/en_US.js:
```diff
@@ -937,6 +937,8 @@ const messages = {
   "workflow.search.help.string15": "property:title=value",
   "workflow.search.help.string16": "property name:",
   "workflow.search.help.string17": "property:name=value",
+  "node.step.plugin.plural": "{0} Node Step Plugins",
```
I does not look like these changes are related

jayas006 replied:
> I does not look like these changes are related

This may be from the conflict files.

mrdubr commented on rundeckapp/grails-spa/packages/ui-trellis/src/library/components/plugins/ChoosePluginModal.vue:
```diff
@@ -17,24 +17,39 @@
-            class="list-group-item"
-            @click.prevent="chooseProviderAdd(service.service, prov.name)"
-            data-testid="provider-button"
+          <template
```
I does not look like these changes are related

mrdubr commented on rundeckapp/grails-spa/packages/ui-trellis/src/library/components/plugins/tests/ChoosePluginModal.spec.ts:
```diff
@@ -1,79 +1,118 @@
-import { shallowMount, VueWrapper, flushPromises } from "@vue/test-utils";
+import { mount, VueWrapper, flushPromises } from "@vue/test-utils";
 import ChoosePluginModal from "../ChoosePluginModal.vue";
-import { Modal, Btn, Tabs, Tab } from "uiv";
```
I does not look like these changes are related

mrdubr commented on rundeckapp/grails-spa/packages/ui-trellis/src/library/stores/Plugins.ts:
```diff
@@ -18,7 +18,7 @@ export class PluginStore {
   async load(service: string): Promise<void> {
     if (this.pluginsByService[service]) return void 0;
     const plugins = await this.client.apiRequest({
-      pathTemplate: "api/40/plugin/list",
```
I does not look like these changes are related

jayas006 commented:
> Please, see my comments.
> 
> Also, please review the changes you're submitting for the review. It appears that some of the changes are unrelated.

Thanks for the feedback. I have removed all unrelated files, corrected variable names, and indented all files properly. Thank you for your patience.

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,75 @@
+        // Set up WebDriverWait to wait up to 30 seconds for elements that take time to appear or change
+        wait = new WebDriverWait(getDriver(), Duration.ofSeconds(30))
+        (go LoginPage).login(TEST_USER, TEST_PASS)
+        def jobDefinition = JobUtils.generateScheduledExecutionXml("TestJobActivityHistory")
```
Do you need to generate and execute a job for every test or can this be done once?
If so, this will result in a serious perf gain.

jayas006 replied:
> Do you need to generate and execute a job for every test or can this be done once? If so, this will result in a serious perf gain.

I have tried fixing it, can you please check if its look ok? Thanks

mrdubr reviewed (changes requested):
The code itself looks great! 👍 
Let's see if we can address the perf issue commented on and we should be good to go.

mrdubr commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -0,0 +1,75 @@
+
+    def setup() {
+        // Set up WebDriverWait to wait up to 30 seconds for elements that take time to appear or change
+        wait = new WebDriverWait(getDriver(), Duration.ofSeconds(30))
```
do you need to do this for every test or just once?

jayas006 commented:
> The code itself looks great! 👍 
> 
> Let's see if we can address the perf issue commented on and we should be good to go.

Thanks Alex,I will fix this so that it's ready for 5.9 release.

jayas006 commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -55,11 +58,10 @@ class JobActivityHistorySpec extends SeleniumBase {
         then: "Validate the saved filter is applied"
         wait.until {
-            jobListPage.getAppliedFilterName()=="TestFilter"
+            jobListPage.getAppliedFilterName() == "TestFilter"
```
The code is auto-formatted by the IDE

mrdubr replied:
Looks good!

mrdubr reviewed (approved):
Looks great!
Love how compact and elegant your final solution is!
👍
